sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1044,6 +1044,34 @@ class Analyzer(
           failOnOuterReference(p)
           p
       }
+
+      // SPARK-17348
+      // Looking for a potential incorrect result case.
+      // When a correlated predicate is a non-equality predicate
+      // it must be placed at the immediate child operator.
+      // Otherwise, the pull up of the correlated predicate
+      // will generate a plan with a different semantics
+      // which could return incorrect result.
+      var continue : Boolean = true
```

Comment:
;-) An explicit question, eh?

DB2 uses a different approach. It does not aggressively transform subquery into join. I will speak of it in Spark terms. DB2 does not perform any query transformation during the Analyzer phase where parts of the LogicalPlan are still unresolved. This is because it doesn't know what the semantics of the entire query is. DB2 does a smalerl scope, comparing to what Spark does today, of transforming subquery to join in (rule-based) Optimizer phase, only with equality correlated predicate and when subquery referencing columns from immediate parents. Therefore DB2 does not encounter the problem we discuss in this PR.

One reason DB2 does differently is DB2 supports subquery processing at run-time so any correlated subquery left from the compiler/optimizer still can be processed.

I will touch on a broader scope of subquery transformation in the next update